geo/maxmind.go

```diff
@@ -0,0 +1,87 @@
+package geo
+
+import (
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/oschwald/geoip2-golang"
+)
+
+const (
+	MaxMindUrl     = "https://raw.githubusercontent.com/leo108/geolite2-db/master/Country.mmdb"
+	MaxMindExpired = 30 * 24 * time.Hour
+)
+
+type MaxMindProvider struct {
+	DB         *geoip2.Reader
+	DBFile     string
+	LastUpdate time.Time
+}
+
+func (p *MaxMindProvider) GetLocation(ip string) (*Location, error) {
+	if time.Since(p.LastUpdate) > MaxMindExpired {
+		err := os.Remove(p.DBFile)
+		if err != nil {
+			return nil, err
+		}
+		tmp, err := NewMaxMindProvider(p.DBFile)
+		if err != nil {
+			return nil, err
+		}
+		p.DB = tmp.DB
+		p.LastUpdate = tmp.LastUpdate
+	}
+	loc, err := p.getLocationFromMM(ip)
+	if err != nil {
+		return &emptyLocation, err
+	}
+	return loc, nil
+}
+
+func NewMaxMindProvider(DBFile string) (*MaxMindProvider, error) {
+	if len(DBFile) == 0 {
+		return nil, errors.New("invalid DBFile path")
+	}
+	if _, err := os.Stat(DBFile); os.IsNotExist(err) {
+		MMFile, err := os.Create(DBFile)
+		if err != nil {
+			return nil, err
+		}
+		defer MMFile.Close()
+
+		resp, err := http.Get(MaxMindUrl)
```

Comment:
Never use http.get, always creates a `http.Client` and set a timeout, otherwise this function may stuck forever. You can use a very long timeout like 30 or 60 minutes.